pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.java

```diff
@@ -75,6 +75,10 @@
             description = "Specifies the string that marks a line which PMD should ignore; default is NOPMD.")
     private String suppressmarker = "NOPMD";
 
+    @Parameter(names = "-multifileanalysisdirectory",
+            description = "Specifies a directory that contains sources the may be used in a multi-file analysis; default is empty.")
```

Comment:
This option is language dependent, I agree with @oowekyala that this would be a fit for language properties.

So that I understand, how this works: The directory, you specify here, contains the file "sfdx-project.json". Where does this file come from? Is it generated by an execution of ApexLink before executing PMD? Is this directory a different directory than you would specify using "-dir" (the "Root directory for sources.")? I'm unfamiliar with the structure of Apex projects.

For me, this looks like a property to enable/make use of ApexLink, rather than enabling multifileanalysis... and ApexLink happens to understand already the relationships between the files that are being analyzed (that's what we want with multifile analysis). PMD already has the knowledge where all files a located (-dir), but doesn't take advantage of this, because we analyze file by file and don't build up a global metadata structure like "sfdx-project.json". Hence my question: Who builds this file and when?

Ok, looking at e.g. https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/master/sfdx-project.json this file actually just contains the path to the sources (possibly multiple paths). So, ApexLink then builds the metadata structure, when creating the Org, right?

I assume, if the paths specified by sfdx-project.json and the paths specified by "-dir" don't match, that would be a configuration error and I'm wondering, what we could do to make it easier for users to call PMD. For apex projects, that have a sfdx-project structure (https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_source_file_format.htm) we could say, that you need to specify the directory where sfdx-project.json is located via "-dir" and PMD would then add the referenced directories, as if specified manually via "-dir". In that case, we might have instead a parameter to opt-in/opt-out for that feature (apex language property: consider directories specified in sfdx-project.json).

Reply:
> Ok, looking at e.g. https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/master/sfdx-project.json this file actually just contains the path to the sources (possibly multiple paths). So, ApexLink then builds the metadata structure, when creating the Org, right?

Yes, it reads the directories from sfdx-project.json to work out what to include in the analysis. The handling of this used to be built into ApexLink but I recently split into a separate library. There is a bit more to the format than being multiple directories, it's really more like multiple modules with either implicit or explicit dependencies between them. Treating them as modules matters in a multi-file analysis because there are rules around visibility and duplicates etc that you can't handle if you think of them as just the sum of a set of directories. There is also .forceignore which is similar in nature to .gitignore, if present its in the same directory as sfdx-project.json.

The sfdx-project.json also contains a namespace for the project. It's critical this is available for multi-file work as it enables the analysis to understand package boundaries. For single file analysis I would imagine a few rules might find it also handy if it could be made available.

> 
> I assume, if the paths specified by sfdx-project.json and the paths specified by "-dir" don't match, that would be a configuration error and I'm wondering, what we could do to make it easier for users to call PMD. For apex projects, that have a sfdx-project structure (https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_source_file_format.htm) we could say, that you need to specify the directory where sfdx-project.json is located via "-dir" and PMD would then add the referenced directories, as if specified manually via "-dir". In that case, we might have instead a parameter to opt-in/opt-out for that feature (apex language property: consider directories specified in sfdx-project.json).

Having "-dir" understand sfdx-project.json makes sense to me as a general approach but likely you want to delegate resolving "-dir" to a set of class files for analysis to a Salesforce specific handler that works alongside the existing handling.

The library I have for this might be a bit overweight for what PMD needs as ApexLink is interested in other project files and they can have some quite unusual handling behaviours. The part I have that is probably most valuable to PMD is the forceignore handling which should be straight forward to port to Java from its current Scala implementation if that made sense.